File: 1574-Shortest Subarray to be Removed to Make Array Sorted.py
class Solution(object):
    def findLengthOfShortestSubarray(self, arr):
        """
        :type arr: List[int]
        :rtype: int
        """
# https://leetcode.com/problems/shortest-subarray-to-be-removed-to-make-array-sorted/discuss/830699/Python-Two-Pointers-Approach-with-Explanation
# 1. l is the last index of prefix such that arr[l+1] < arr[l]
# 2. r is the first index of suffix such that arr[r-1] > arr[r]
# 3. Try to "merge 2 sorted arrays"

        l, r = 0, len(arr)-1
        
        while l < r:
            if arr[l] > arr[l+1]:
                break
            l += 1
            
        if l == r:
            return 0
        
        while r > l:
            if arr[r] < arr[r-1]:
                break
            r -= 1
            
        toRemove = min(len(arr)-l-1, r)
        
        for idx in range(l+1):
            if arr[idx] <= arr[r]:
                toRemove = min(toRemove, r-idx-1)
            elif r < len(arr) - 1:
                r += 1
            else:
                break
        
        return toRemove

File: C++/1574-Shortest Subarray to be Removed to Make Array Sorted.cpp
// First AC, binary search
class Solution {
public:
    int findLengthOfShortestSubarray(vector<int>& arr) {
        int N = arr.size(), res = 0;
        vector<int> left(N+1, INT_MAX);
        left[0] = 0;

        for (int i=0; i<N; i++) {
            if (i == 0) {
                left[i+1] = arr[i];
                res++;
            } else {
                if (arr[i] >= arr[i-1]) {
                    left[i+1] = arr[i];
                    res++;
                } else {
                    break;
                }
            }
        }

        for (int i=N-1; i>=0; i--) {
            if (i == N-1 || arr[i] <= arr[i+1]) {
                int idx = upper_bound(left.begin(), left.begin()+i, arr[i]) - left.begin();
                idx--;
                res = max(res, N-i+idx);
            } else {
                break;
            }
        }

        return N-res;
    }
};



// Efficient two-pointer solution
class Solution {
public:
    int findLengthOfShortestSubarray(vector<int>& arr) {
        int N = arr.size(), res = N, left = 0, right = N-1;
        while (left+1<N && arr[left] <= arr[left+1]) left++;
        if (left == N-1) return 0;
        while (right > 0 && arr[right] >= arr[right-1]) right--;
        res = min(res, N-max(left+1, N-right));
        
        for (int i=0; i<=left; i++) {
            if (arr[i] <= arr[right]) {
                res = min(res, right-i-1);
            } else if (right < N-1){
                right++;
            } else {
                break;
            }
        }

        return res;
    }
};